tmt/steps/report/junit.py

```diff
@@ -58,40 +42,170 @@ def duration_to_seconds(duration: Optional[str]) -> Optional[int]:
+        template_path: Optional[Path] = None,
+        results_context: Optional[ResultsContext] = None,
+        **extra_variables: Any
+        ) -> str:
+    """
+    Create JUnit XML file and return the data
 
-    suite = junit_xml.TestSuite(report.step.plan.name)
+    :param report: instance of a ReportPlugin.
+    :param flavor: name of a JUnit flavor to generate.
+    :param template_path: if set, the provided template will be used instead of
+        a pre-defined flavor template. In this case, the ``flavor`` must be set
+        to ``custom`` value.
+    :pawam extra_variables: if set, these variables get propagated into the
+        Jinja template.
+    """
+
+    # Get the template context for TMT results
+    results_context = results_context or ResultsContext(report.step.plan.execute.results())
 
-    for result in report.step.plan.execute.results():
+    # Prepare the template environment
+    def _read_log(log: Path) -> str:
+        """ Read the contents of a given result log """
         try:
-            main_log = report.step.plan.execute.read(result.log[0])
-        except (IndexError, AttributeError):
-            main_log = None
-        case = junit_xml.TestCase(
-            result.name,
-            classname=None,
-            elapsed_sec=duration_to_seconds(result.duration))
-
-        if report.data.include_output_log:
-            case.stdout = main_log
-
-        # Map tmt OUTCOME to JUnit states
-        if result.result == tmt.result.ResultOutcome.ERROR:
-            case.add_error_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.FAIL:
-            case.add_failure_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.INFO:
-            case.add_skipped_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.WARN:
-            case.add_error_info(result.result.value, output=result.failures(main_log))
-        # Passed state is the default
-        suite.test_cases.append(case)
-
-    return suite
+            return str(report.step.plan.execute.read(log))
+        except AttributeError:
+            return ''
+
+    environment = default_template_environment()
+
+    # Use a FileSystemLoader for a non-custom flavor
+    if flavor != CUSTOM_FLAVOR_NAME:
+        environment.loader = FileSystemLoader(searchpath=Path('tmt/steps/report/junit/templates/'))
+        template_path = Path(tmt.utils.resource_files(
+            Path(f'steps/report/junit/templates/{flavor}.xml.j2')))
+
+    environment.filters.update({
+        'read_log': _read_log,
+        'duration_to_seconds': duration_to_seconds,
+        })
+
+    # Explicitly enable the autoescape because it's disabled by default by TMT
+    # (see /teemtee/tmt/issues/2873 for more info.
+    environment.autoescape = select_autoescape(enabled_extensions=('xml'))
```

Comment:
Is `xml` correct, or does Jinja cut away the `.j2` suffix?

Reply:
I think you might be right:
- https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.select_autoescape

```
The select_autoescape() function returns a function that works roughly like this:

def autoescape(template_name):
    if template_name is None:
        return False
    if template_name.endswith(('.html', '.htm', '.xml'))
```

I will change it to `xml.j2`